File: sdks/go/datalens/client_test.go
package datalens

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func TestDecodedEventsConnectionSendsBearerTokenAndGraphQLBody(t *testing.T) {
	var captured struct {
		Path          string
		Authorization string
		Application   string
		UserAgent     string
		Body          graphqlRequest
	}
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		captured.Path = r.URL.Path
		captured.Authorization = r.Header.Get("Authorization")
		captured.Application = r.Header.Get(ApplicationHeader)
		captured.UserAgent = r.Header.Get("User-Agent")
		if err := json.NewDecoder(r.Body).Decode(&captured.Body); err != nil {
			t.Fatalf("decode request body: %v", err)
		}
		respondJSON(t, w, http.StatusOK, map[string]any{
			"data": map[string]any{
				"decodedEventsConnection": map[string]any{
					"nodes": []any{
						map[string]any{
							"dataset":         "evm.logs",
							"blockNumber":     12,
							"eventName":       "MessageSent",
							"transactionHash": "0xabc",
							"decodedArgs":     map[string]any{"sender": "0x1"},
							"payload":         map[string]any{"raw": true},
						},
					},
					"edges": []any{},
					"pageInfo": map[string]any{
						"endCursor":   "cursor-1",
						"hasNextPage": false,
					},
				},
			},
		})
	}))
	defer server.Close()

	client, err := NewClient(server.URL,
		WithToken("secret-token"),
		WithApplication("query-app"),
		WithUserAgent("datalens-test/1.0"),
	)
	if err != nil {
		t.Fatalf("new client: %v", err)
	}

	page, err := client.DecodedEventsConnection(context.Background(), EventFilter{
		Dataset:   "evm.logs",
		EventName: "MessageSent",
		First:     25,
		After:     "before-cursor",
	})
	if err != nil {
		t.Fatalf("decoded events connection: %v", err)
	}

	if captured.Path != "/index/graphql" {
		t.Fatalf("path = %q, want /index/graphql", captured.Path)
	}
	if captured.Authorization != "Bearer secret-token" {
		t.Fatalf("authorization = %q", captured.Authorization)
	}
	if captured.Application != "query-app" {
		t.Fatalf("application = %q", captured.Application)
	}
	if captured.UserAgent != "datalens-test/1.0" {
		t.Fatalf("user agent = %q", captured.UserAgent)
	}
	if !strings.Contains(captured.Body.Query, "decodedEventsConnection") {
		t.Fatalf("query did not request decodedEventsConnection: %s", captured.Body.Query)
	}
	variables := captured.Body.Variables
	assertEqual(t, variables["dataset"], "evm.logs")
	assertEqual(t, variables["eventName"], "MessageSent")
	assertEqual[any](t, variables["first"], float64(25))
	assertEqual(t, variables["after"], "before-cursor")
	if len(page.Nodes) != 1 {
		t.Fatalf("decoded nodes length = %d", len(page.Nodes))
	}
	assertEqual(t, page.Nodes[0].EventName, "MessageSent")
	assertEqual(t, page.PageInfo.EndCursor, "cursor-1")
}

func TestFetchAllRawEventsPaginatesWithEndCursor(t *testing.T) {
	var afterValues []any
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		var body graphqlRequest
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatalf("decode request body: %v", err)
		}
		afterValues = append(afterValues, body.Variables["after"])
		cursor := "cursor-1"
		hasNext := len(afterValues) == 1
		block := 1
		if !hasNext {
			cursor = "cursor-2"
			block = 2
		}
		respondJSON(t, w, http.StatusOK, map[string]any{
			"data": map[string]any{
				"eventsConnection": map[string]any{
					"nodes": []any{
						map[string]any{
							"dataset":         "evm.logs",
							"blockNumber":     block,
							"transactionHash": "0xhash",
							"topics":          []any{"0xtopic"},
							"decoded":         map[string]any{},
							"payload":         map[string]any{},
						},
					},
					"edges": []any{},
					"pageInfo": map[string]any{
						"endCursor":   cursor,
						"hasNextPage": hasNext,
					},
				},
			},
		})
	}))
	defer server.Close()

	client, err := NewClient(server.URL)
	if err != nil {
		t.Fatalf("new client: %v", err)
	}

	events, err := client.FetchAllRawEvents(context.Background(), EventFilter{Dataset: "evm.logs", First: 1})
	if err != nil {
		t.Fatalf("fetch all raw events: %v", err)
	}

	if len(events) != 2 {
		t.Fatalf("events length = %d", len(events))
	}
	if afterValues[0] != nil {
		t.Fatalf("first after = %#v, want nil", afterValues[0])
	}
	assertEqual(t, afterValues[1], "cursor-1")
	assertEqual(t, events[1].BlockNumber, 2)
}

func TestNativeDiscoveryUsesNativeGraphQLEndpoint(t *testing.T) {
	var captured graphqlRequest
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/native/graphql" {
			t.Fatalf("path = %q, want /native/graphql", r.URL.Path)
		}
		if err := json.NewDecoder(r.Body).Decode(&captured); err != nil {
			t.Fatalf("decode request body: %v", err)
		}
		respondJSON(t, w, http.StatusOK, map[string]any{
			"data": map[string]any{
				"discovery": map[string]any{
					"chains": []any{
						map[string]any{
							"identity": map[string]any{"configuredName": "ethereum"},
							"datasets": []any{
								map[string]any{
									"datasetKey": "evm.logs",
									"rangeKinds": map[string]any{"block": true},
									"selectors":  []any{"all", "evm_logs"},
									"enabled":    true,
								},
							},
						},
					},
				},
			},
		})
	}))
	defer server.Close()

	client, err := NewClient(server.URL)
	if err != nil {
		t.Fatalf("new client: %v", err)
	}

	discovery, err := client.Discovery(context.Background())
	if err != nil {
		t.Fatalf("discovery: %v", err)
	}

	if !strings.Contains(captured.Query, "discovery") {
		t.Fatalf("query did not request discovery: %s", captured.Query)
	}
	if len(discovery.Chains) != 1 {
		t.Fatalf("chains length = %d", len(discovery.Chains))
	}
	assertEqual(t, discovery.Chains[0].Datasets[0].DatasetKey, "evm.logs")
}

func TestGraphQLErrorsExposeStableKind(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		respondJSON(t, w, http.StatusOK, map[string]any{
			"errors": []any{
				map[string]any{
					"message": "unsupported dataset",
					"extensions": map[string]any{
						"code": "UnsupportedDataset",
						"kind": "unsupported_dataset",
					},
				},
			},
		})
	}))
	defer server.Close()

	client, err := NewClient(server.URL)
	if err != nil {
		t.Fatalf("new client: %v", err)
	}

	_, err = client.RawEventsConnection(context.Background(), EventFilter{Dataset: "evm.blocks"})
	if err == nil {
		t.Fatal("expected error")
	}
	var gqlErr *GraphQLError
	if !errors.As(err, &gqlErr) {
		t.Fatalf("error type = %T, want *GraphQLError", err)
	}
	assertEqual(t, gqlErr.Errors[0].Kind, "unsupported_dataset")
	assertEqual(t, gqlErr.Errors[0].Code, "UnsupportedDataset")
}

func TestHTTPAuthAndRateLimitErrorsUseTypedErrors(t *testing.T) {
	statuses := []int{http.StatusInternalServerError, http.StatusUnauthorized, http.StatusTooManyRequests}
	for _, status := range statuses {
		t.Run(http.StatusText(status), func(t *testing.T) {
			server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
				respondJSON(t, w, status, map[string]any{"error": map[string]any{"kind": "rate_limited", "message": "limited"}})
			}))
			defer server.Close()

			client, err := NewClient(server.URL)
			if err != nil {
				t.Fatalf("new client: %v", err)
			}
			_, err = client.RawEventsConnection(context.Background(), EventFilter{Dataset: "evm.logs"})
			if err == nil {
				t.Fatal("expected error")
			}
			switch status {
			case http.StatusInternalServerError:
				var httpErr *HTTPError
				if !errors.As(err, &httpErr) {
					t.Fatalf("error type = %T, want *HTTPError", err)
				}
			case http.StatusUnauthorized:
				var authErr *AuthError
				if !errors.As(err, &authErr) {
					t.Fatalf("error type = %T, want *AuthError", err)
				}
			case http.StatusTooManyRequests:
				var rateErr *RateLimitError
				if !errors.As(err, &rateErr) {
					t.Fatalf("error type = %T, want *RateLimitError", err)
				}
			}
		})
	}
}

func respondJSON(t *testing.T, w http.ResponseWriter, status int, body any) {
	t.Helper()
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	if err := json.NewEncoder(w).Encode(body); err != nil {
		t.Fatalf("encode response: %v", err)
	}
}

func assertEqual[T comparable](t *testing.T, got, want T) {
	t.Helper()
	if got != want {
		t.Fatalf("got %#v, want %#v", got, want)
	}
}

File: sdks/typescript/test/client.test.ts
import assert from "node:assert/strict";
import { createServer, type IncomingMessage, type ServerResponse } from "node:http";
import { after, before, test } from "node:test";

import {
  DatalensAuthError,
  DatalensClient,
  DatalensGraphQLError,
  DatalensHttpError,
  DatalensRateLimitError,
  type JsonValue,
} from "../src/index.js";

type RequestRecord = {
  body: {
    query?: string;
    variables?: Record<string, unknown>;
    operationName?: string;
  };
  headers: IncomingMessage["headers"];
  path: string;
};

const requests: RequestRecord[] = [];

let nextResponse:
  | {
      status?: number;
      headers?: Record<string, string>;
      body: unknown;
    }
  | undefined;
let queuedResponses: NonNullable<typeof nextResponse>[] = [];

const server = createServer(async (request: IncomingMessage, response: ServerResponse) => {
  const chunks: Buffer[] = [];
  for await (const chunk of request) {
    chunks.push(Buffer.from(chunk));
  }

  const body = JSON.parse(Buffer.concat(chunks).toString("utf8"));
  requests.push({
    body,
    headers: request.headers,
    path: request.url ?? "",
  });

  const planned = queuedResponses.shift() ??
    nextResponse ?? {
    body: {
      data: {},
    },
  };
  nextResponse = undefined;

  response.statusCode = planned.status ?? 200;
  response.setHeader("content-type", "application/json");
  for (const [name, value] of Object.entries(planned.headers ?? {})) {
    response.setHeader(name, value);
  }
  response.end(JSON.stringify(planned.body));
});

let endpoint = "";

before(async () => {
  await new Promise<void>((resolve) => {
    server.listen(0, "127.0.0.1", resolve);
  });
  const address = server.address();
  assert(address && typeof address === "object");
  endpoint = `http://127.0.0.1:${address.port}`;
});

after(async () => {
  await new Promise<void>((resolve, reject) => {
    server.close((error) => (error ? reject(error) : resolve()));
  });
});

function resetServer(response: typeof nextResponse): void {
  requests.length = 0;
  queuedResponses = [];
  nextResponse = response;
}

test("queryDecodedEvents sends bearer auth and index GraphQL request shape", async () => {
  resetServer({
    body: {
      data: {
        decodedEventsConnection: {
          edges: [
            {
              cursor: "0",
              node: {
                blockNumber: 10,
                decodedArgs: { msgHash: "0xabc" },
                eventName: "MessageAccepted",
                logIndex: 0,
              },
            },
          ],
          nodes: [
            {
              blockNumber: 10,
              decodedArgs: { msgHash: "0xabc" },
              eventName: "MessageAccepted",
              logIndex: 0,
            },
          ],
          pageInfo: {
            endCursor: "0",
            hasNextPage: false,
          },
        },
      },
    },
  });

  const client = new DatalensClient({
    endpoint,
    token: "secret-token",
    application: "query-app",
  });

  const page = await client.index.queryDecodedEvents({
    dataset: "evm.logs",
    eventName: "MessageAccepted",
    first: 25,
  });

  assert.equal(requests[0].path, "/index/graphql");
  assert.equal(requests[0].headers.authorization, "Bearer secret-token");
  assert.equal(requests[0].headers["x-datalens-application"], "query-app");
  assert.match(requests[0].headers["user-agent"] as string, /^datalens-typescript-sdk\//);
  assert.match(requests[0].body.query ?? "", /decodedEventsConnection/);
  assert.deepEqual(requests[0].body.variables, {
    dataset: "evm.logs",
    eventName: "MessageAccepted",
    first: 25,
  });
  assert.equal(page.nodes[0].decodedArgs.msgHash, "0xabc");
});

test("queryRawEvents uses index GraphQL eventsConnection", async () => {
  resetServer({
    body: {
      data: {
        eventsConnection: {
          edges: [
            {
              cursor: "0",
              node: {
                blockNumber: 20,
                eventIndex: 1,
                payload: { raw: true },
                topics: ["0xtopic"],
              },
            },
          ],
          nodes: [
            {
              blockNumber: 20,
              eventIndex: 1,
              payload: { raw: true },
              topics: ["0xtopic"],
            },
          ],
          pageInfo: {
            endCursor: "0",
            hasNextPage: false,
          },
        },
      },
    },
  });

  const client = new DatalensClient({ endpoint });
  const page = await client.index.queryRawEvents({
    dataset: "evm.logs",
    first: 10,
    topic0: "0xtopic",
  });

  assert.equal(requests[0].path, "/index/graphql");
  assert.match(requests[0].body.query ?? "", /eventsConnection/);
  assert.equal((page.nodes[0].payload as { raw: JsonValue }).raw, true);
});

test("paginate follows cursor pageInfo until the connection is exhausted", async () => {
  const pages = [
    {
      data: {
        decodedEventsConnection: {
          edges: [{ cursor: "0", node: { blockNumber: 10, decodedArgs: {} } }],
          nodes: [{ blockNumber: 10, decodedArgs: {} }],
          pageInfo: { endCursor: "0", hasNextPage: true },
        },
      },
    },
    {
      data: {
        decodedEventsConnection: {
          edges: [{ cursor: "1", node: { blockNumber: 20, decodedArgs: {} } }],
          nodes: [{ blockNumber: 20, decodedArgs: {} }],
          pageInfo: { endCursor: "1", hasNextPage: false },
        },
      },
    },
  ];

  requests.length = 0;
  queuedResponses = pages.map((body) => ({ body }));
  nextResponse = undefined;

  const client = new DatalensClient({ endpoint });
  const blocks: number[] = [];

  for await (const event of client.index.paginateDecodedEvents({
    dataset: "evm.logs",
    first: 1,
  })) {
    blocks.push(event.blockNumber ?? 0);
  }

  assert.deepEqual(blocks, [10, 20]);
  assert.deepEqual(requests.map((request) => request.body.variables), [
    { dataset: "evm.logs", first: 1 },
    { dataset: "evm.logs", first: 1, after: "0" },
  ]);
});

test("native query and discovery use native GraphQL operations", async () => {
  resetServer({
    body: {
      data: {
        query: {
          chain: { configuredName: "ethereum" },
          datasetKey: "evm.blocks",
          range: { kind: "block", start: 10, end: 10 },
          cache: { hit_ranges: [] },
          rows: { rows: [{ number: 10 }] },
        },
      },
    },
  });

  const client = new DatalensClient({ endpoint });
  const query = await client.native.query({
    chain: {
      family: { kind: "evm" },
      configuredName: "ethereum",
      networkId: { numeric: 1 },
    },
    datasetKey: { family: "evm", name: "blocks" },
    selector: { kind: "all" },
    range: { kind: "block", start: 10, end: 10 },
    finality: "durable_only",
    fields: {},
  });

  assert.equal(requests[0].path, "/native/graphql");
  assert.match(requests[0].body.query ?? "", /query\(input: \$input\)/);
  assert.equal(query.datasetKey, "evm.blocks");

  resetServer({
    body: {
      data: {
        discovery: {
          chains: [
            {
              identity: { configuredName: "ethereum" },
              datasets: [
                {
                  datasetKey: "evm.blocks",
                  enabled: true,
                  rangeKinds: [{ kind: "block" }],
                  selectors: ["all"],
                },
              ],
            },
          ],
        },
      },
    },
  });

  const discovery = await client.native.discovery();
  assert.match(requests[0].body.query ?? "", /discovery/);
  assert.equal(discovery.chains[0].datasets[0].datasetKey, "evm.blocks");
});

test("GraphQL errors are exposed as stable SDK errors", async () => {
  resetServer({
    body: {
      errors: [
        {
          message: "rate limit exceeded",
          extensions: {
            code: "RATE_LIMITED",
            kind: "RateLimited",
          },
        },
      ],
      data: null,
    },
  });

  const client = new DatalensClient({ endpoint });

  await assert.rejects(
    client.index.queryRawEvents({ dataset: "evm.logs" }),
    (error: unknown) =>
      error instanceof DatalensGraphQLError &&
      error instanceof DatalensRateLimitError &&
      error.kind === "RateLimited",
  );
});

test("GraphQL auth errors are exposed as auth errors", async () => {
  resetServer({
    body: {
      errors: [
        {
          message: "invalid credentials",
          extensions: {
            code: "AUTHENTICATION_FAILED",
            kind: "AuthenticationFailed",
          },
        },
      ],
      data: null,
    },
  });

  const client = new DatalensClient({ endpoint });

  await assert.rejects(
    client.native.discovery(),
    (error: unknown) =>
      error instanceof DatalensGraphQLError &&
      error instanceof DatalensAuthError &&
      error.kind === "AuthenticationFailed",
  );
});

test("HTTP auth and rate-limit responses use specialized errors", async () => {
  const client = new DatalensClient({ endpoint });

  resetServer({
    status: 401,
    body: { error: { kind: "unauthorized", message: "missing token" } },
  });
  await assert.rejects(
    client.native.chains(),
    (error: unknown) => error instanceof DatalensAuthError && error.status === 401,
  );

  resetServer({
    status: 429,
    headers: { "retry-after": "30" },
    body: { error: { kind: "rate_limited", message: "too many requests" } },
  });
  await assert.rejects(
    client.native.chains(),
    (error: unknown) =>
      error instanceof DatalensRateLimitError &&
      error instanceof DatalensHttpError &&
      (error as DatalensRateLimitError).retryAfter === "30",
  );
});
